zrml/swaps/src/tests.rs

```diff
@@ -1655,6 +1759,107 @@ fn create_pool_fails_on_total_weight_above_maximum_total_weight() {
     });
 }
 
+#[test]
+fn pool_join_fails_if_max_assets_in_is_violated() {
+    ExtBuilder::default().build().execute_with(|| {
+        create_initial_pool_with_funds_for_alice(ScoringRule::CPMM, true);
+        assert_noop!(
+            Swaps::pool_join(alice_signed(), 0, _1, vec!(_1, _1, _1 - 1, _1)),
+            crate::Error::<Runtime>::LimitIn,
+        );
+    });
+}
+
+#[test]
+fn pool_join_with_exact_asset_amount_fails_if_min_pool_tokens_is_violated() {
+    ExtBuilder::default().build().execute_with(|| {
+        create_initial_pool_with_funds_for_alice(ScoringRule::CPMM, true);
+        assert_noop!(
+            Swaps::pool_join_with_exact_asset_amount(
+                alice_signed(),
+                0,
+                ASSET_A,
+                _1,
+                2490679300 + 100,
+            ),
+            crate::Error::<Runtime>::LimitOut,
+        );
+    });
+}
+
+#[test]
+fn pool_join_with_exact_pool_amount_fails_if_max_asset_amount_is_violated() {
+    ExtBuilder::default().build().execute_with(|| {
+        create_initial_pool_with_funds_for_alice(ScoringRule::CPMM, true);
+        assert_noop!(
+            Swaps::pool_join_with_exact_pool_amount(
+                alice_signed(),
+                0,
+                ASSET_A,
+                _1,
+                40604010000 - 100,
+            ),
+            crate::Error::<Runtime>::LimitIn,
+        );
+    });
+}
+
+#[test]
+fn pool_exit_fails_if_min_assets_out_is_violated() {
+    ExtBuilder::default().build().execute_with(|| {
+        create_initial_pool_with_funds_for_alice(ScoringRule::CPMM, true);
+        assert_ok!(Swaps::pool_join(alice_signed(), 0, _1, vec!(_1, _1, _1, _1)));
+        assert_noop!(
+            Swaps::pool_exit(alice_signed(), 0, _1, vec!(_1, _1, _1 + 1, _1)),
```

Comment:
mentioned above (could be any asset bound)

Reply:
I think that adding `1` is a good indicator that we're just going "a bit too far". Think of it this way: We just joined the pool 1:1 and now to to get a little more out than we put in. This should fail.

Reply:
In fact, `1` is not an arbitrary number in this case.